### ISSUE TITLE
Sweep for auxillary loss and top-k

### ISSUE THREAD

**Issue by aaronkl** [OpenEuroLLM/Taskboard#165]
Sweep over auxiliary loss and top-k for the 200M-A50M model.  Also try it with different number of experts

**Comment by swag2198** [comment]
## Experiment details
- Starting config: Aaron's ~200M total with ~50M active with 8 total experts and 2 activated experts (topk)
- Learning rate and batch size fixed to 0.007 and 512 (informed by #162 although the configs are different)
- I mainly sweep over number of experts (8 to 128) and aux loss strength. Also use 2 topk: 2 and 4 to see variations.
- The runs can be found [here (wandb)](https://wandb.ai/openeurollm-project/moe-ablations).
- Configs for the sweep here: [github](https://github.com/OpenEuroLLM/oellm-autoexp/tree/exp_swagatam/config/experiments/swagatam/moe_ablations_small).

I discuss and interpret some results below.

## 1. Validation loss vs. MoE sparsity
- x-axis: number of experts are varies from 8 to 128, while keeping activated experts fixed (2 and 4 here).
- y-axis: final validation loss (left) and load balancing loss (LBL) or aux. loss (right) after training with 10B tokens.
- **IsoFLOPs**: Since the activated experts are fixed for each curve and we always train for 10B tokens, each sparsity setting on a curve expends constant (theoretical) FLOPs, so sparsity helps to buy performance (lower validation loss) at the cost of more total number of parameters. Aligns with literature, also not surprising that higher activated experts (more computational budget) helps.
- What's surprising to me is that the load balancing loss also constantly decreases with sparsity (higher experts), atleast for the topk=2 case! I thought in principle having more experts (and selecting only few, 2 or 4 of them) would make it difficult for the router to balance the load among them, so that would reflect in aux loss being high. This may imply aux loss values might not correlate with degree of difficulty of load balancing (not sure)! Also the load balancing loss is higher for 4 active experts.
<img width="1775" height="872" alt="Image" src="https://github.com/user-attachments/assets/f85136ea-d39b-4265-a14f-53125bf99624" />


## 2. Effect of aux loss coefficient strength ($\alpha$)
- We sweep over a range of $\alpha$ values, the default being $1e-3$ and plot the final validation loss and aux loss values after training with 10B tokens.
- We here run it for 16 and 64 total experts, keeping the topk fixed to 2 (so both curves are IsoFLOPs but have different sparsity/total parameters). Note that the validation loss curve for 64 experts is always below the same for 16 experts, so sparsity helps again for both good and bad load balancing.
- At very low strength, the load balancing essentially does not work and the validation losses are also poor (higher). Note that here, we indeed see higher aux loss for 64 experts for $\alpha=1e-6$, but on increasing $\alpha$, the configs quickly converge to a lower aux loss value ~ 1.
- From the val loss plot, looks like there is a optimal choice of $\alpha$ here where we get maximum "benefit from aux loss" (regularization?) as the lm loss is lowest. Increasing $\alpha > 5e-3$ degrades the model.
<img width="1777" height="868" alt="Image" src="https://github.com/user-attachments/assets/026672a8-8859-48a3-bcfe-78e9b0734f4a" />

## Next steps
- Decide on a larger model config?
- Also design learning rate and batch size grid (following @Niccolo-Ajroldi's dense scaling laws grid)
- Track evals and integrate @abhash-er's metrics #164 as basically another health check during pretraining
